avocado/core/test.py

```diff
@@ -133,6 +133,98 @@ def str_filesystem(self):
                                  % (self.str_uid, str(self)))
 
 
+def get_file_data_dir(test):
+    """
+    Returns the path to the file-level data directory
+
+    For test a test file hosted at /usr/share/avocado/tests/sleeptest.py
+    the file-level data dir is /usr/share/avocado/tests/sleeptest.py.data.
+
+    Note that this directory has no specific relation to the test
+    name, only to the file that contains the test.  It can be used to
+    host data files that are generic enough to be used for all tests
+    contained in a given test file.
+
+    This function is currently a thin wrapper (almost an alias) for
+    :attr:`avocado.core.test.Test.datadir`.  The idea is just to give
+    consistent naming across the various data dir, and prepare for the
+    removal of :attr:`avocado.core.test.Test.datadir`.
+
+    :param test: the test to locate its file-level data directory
+    """
+    # TODO: on removal of Test.datadir, move its code over here
+    return test.datadir
+
+
+def get_test_data_dir(test):
+    """
+    Returns the path to the test-level data directory
+
+    For an INSTRUMENTED test, implemented by a method named `test`, on
+    a class named `SleepTest`, hosted at a file located at
+    /usr/share/avocado/tests/sleeptest.py, the test-level data directory
+    is /usr/share/avocado/tests/sleeptest.py.data/SleepTest.test
+
+    For a SIMPLE test, the test is the file, so this is the same as
+    :func:`get_file_data_dir`.
+
+    :param test: the test to locate its file-level data directory
+    """
+    file_data_dir = get_file_data_dir(test)
+    if file_data_dir is None:
+        return None
+
+    if isinstance(test, SimpleTest):
+        return file_data_dir
+
+    test_level_data_dir = "%s.%s" % (test.__class__.__name__,
+                                     test._testMethodName)
+    test_level_data_dir = astring.string_to_safe_path(test_level_data_dir)
+    return os.path.join(file_data_dir, test_level_data_dir)
+
+
+def get_variant_data_dir(test):
+    """
+    Returns the path to the variant-level data directory
+
+    For an INSTRUMENTED test, implemented by a method named `test`, on
+    a class named `SleepTest`, hosted at a file located at
+    /usr/share/avocado/tests/sleeptest.py, and with variant name
+    "foo-ffff", the variant-level data directory is
+    /usr/share/avocado/tests/sleeptest.py.data/SleepTest.test;foo-ffff
+
+    :param test: the test to locate its file-level data directory
+    """
+    test_data_dir = get_test_data_dir(test)
+    if test_data_dir is None or test.name.variant is None:
+        return None
+
+    variant_dirname = astring.string_to_safe_path(test.name.variant)
```

Comment:
This is actually quite dangerous. On long names many variants can point to the same location. I'd suggest adding a `astring.string_to_safe_path_with_hash` function, which would work as `string_to_safe_path` only in case it needs to shorten the name it should remove additional 4 (or so) characters and hash the original content. Example:

```python
str1 = "foo-bar-baz"
str2 = "foo-bar-baz-123"
string_to_safe_path1 = "foo-bar"
string_to_safe_path2 = "foo-bar"
string_to_safe_path_with_hash1 = "foo4c43"
string_to_safe_path_with_hash2 = "foo2afb"
```

Reply:
Humm... OK, I'll take a look at this.